From: Colin Ian King <colin.king@canonical.com>
Date: Tue, 27 Mar 2018 14:26:01 +0100
Subject: libata: ensure host is free'd on error exit paths
Patch-mainline: v4.17-rc1
Git-commit: dafd6c496381c1cd1f5ba9ad953e810bdcc931bc
References: git-fixes

The host structure is not being kfree'd on two error exit paths
leading to memory leaks. Add in new err_free label and kfree host.

Detected by CoverityScan, CID#1466103 ("Resource leak")

Fixes: 2623c7a5f279 ("libata: add refcounting to ata_host")
Signed-off-by: Colin Ian King <colin.king@canonical.com>
Signed-off-by: Tejun Heo <tj@kernel.org>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 drivers/ata/libata-core.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

--- a/drivers/ata/libata-core.c
+++ b/drivers/ata/libata-core.c
@@ -6075,7 +6075,7 @@ struct ata_host *ata_host_alloc(struct d
 	DPRINTK("ENTER\n");
 
 	if (!devres_open_group(dev, NULL, GFP_KERNEL))
-		return NULL;
+		goto err_free;
 
 	/* alloc a container for our list of ATA ports (buses) */
 	sz = sizeof(struct ata_host) + (max_ports + 1) * sizeof(void *);
@@ -6109,6 +6109,8 @@ struct ata_host *ata_host_alloc(struct d
 
  err_out:
 	devres_release_group(dev, NULL);
+ err_free:
+	kfree(host);
 	return NULL;
 }
 
